CASSANDRA-20827 Represent complex settings as JSON on system_views.settings table

mrktsm opened the pull request:
**JIRA:** https://issues.apache.org/jira/browse/CASSANDRA-20827

Replace Collections.toString() with JSON serialization for arrays, maps, and collections in SettingsTable.getValue() to improve programmatic access.

Complex configuration values were previously displayed using Collections.toString() format:
```
seed_provider.parameters | {seeds=127.0.0.1:7000}
```

Now output as proper JSON for better programmatic parsing:
```
seed_provider.parameters | {"seeds": "127.0.0.1:7000"}
```

**Changes:**
- Modified SettingsTable.getValue() to use JsonUtils.JSON_OBJECT_MAPPER for complex types
- Added exception handling with fallback to toString() for robustness  
- Updated all affected tests to expect JSON array format: ["item1", "item2"] vs [item1, item2]
- Covers data_file_directories, cipher_suites, and protocol configuration tests

**Testing:**
- All existing tests updated and passing with new JSON format
- Verified proper JSON serialization for arrays, maps, and collections
- Exception handling tested via fallback mechanism

patch by Marko Tsymbaliuk for CASSANDRA-20827

mrktsm committed:
Represent complex settings as JSON on system_views.settings table

Replace Collections.toString() with JSON serialization for arrays, maps, and collections
in SettingsTable.getValue() to improve programmatic access. Added fallback to toString()
for error handling and updated all affected tests.

patch by Marko Tsymbaliuk for CASSANDRA-20827

Maxwell-Guo commented on src/java/org/apache/cassandra/db/virtual/SettingsTable.java:
```diff
@@ -93,8 +95,17 @@ String getValue(Property prop)
-            return Arrays.asList((Object[]) value).toString();
+        if (value.getClass().isArray() || value instanceof Collection || value instanceof Map)
+        {
+            try {
```
```
if (value.getClass.isArray())
  return Arrays.asList((Object[]) value).toString();
if (value instanceof Collection || value instanceof Map)
  xxxx
```

mrktsm replied:
`Arrays.asList().toString()` gives `[item1, item2]`, and we need JSON `["item1", "item2"]`?

Maxwell-Guo commented on test/unit/org/apache/cassandra/db/virtual/SettingsTableTest.java:
```diff
@@ -71,7 +71,7 @@ public void config()
     public void testArray() throws Throwable
     {
```
I think we should add more test case here , the value type can be map, set,list , for example : 
  ```
config.seed_provider =  new ParameterizedClass();
        config.seed_provider.class_name = "org.apache.cassandra.locator.SimpleSeedProvider";
        config.seed_provider.parameters = Map.of("seeds", "127.0.0.1:7000");
```
and 
`SELECT * FROM vts.settings`
to see if the map value can be convert to json.

mrktsm replied:
@Maxwell-Guo Of course! That's a great suggestion! I've added the Map test case you requested. 

The new `testMapSerialization()` test uses the exact scenario you mentioned - `seed_provider.parameters` with a Map containing seeds and timeout values. The test verifies that Maps get properly serialized to JSON format.

All tests are passing and the Map serialization is working as expected.

mrktsm committed:
Update SettingsTableTest to match JSON output format

Fix test expectations to match Jackson JSON serializer output without spaces after commas.

mrktsm committed:
Add Map serialization test case for seed_provider.parameters

Addresses reviewer feedback to test Map serialization with real-world example
from the JIRA issue description. Verifies JSON output for complex Map structures.

Maxwell-Guo commented:
Hang on please , I feel like we need to make a decision on the final implementation with daivd and paulo.

mrktsm committed:
Implement JSON serialization with proper credential redaction

Apply Stefan's feedback to fix CASSANDRA-20856 redaction compatibility:

- Extract JSON serialization logic into tryConstructJson() helper method
- Apply redaction to Maps before JSON serialization, not after
- Change redacted Map output from toString() to clean JSON format
- Maintain existing password redaction for keys ending with '_password' or 'password'

This ensures complex settings like seed_provider.parameters display as proper
JSON while still redacting sensitive credentials.

Addresses reviewer feedback from Stefan Miklosovic on CASSANDRA-20827.

mrktsm committed:
Add configurable JSON formatting for complex settings

Implement system property to control output format for complex settings
in system_views.settings table per Paulo Motta and Stefan Miklosovic feedback:

- Add VIRTUAL_TABLE_COMPLEX_SETTINGS_FORMAT_JSON property (default: true)
- Property: cassandra.virtual_table_complex_settings_format_json
- When true: Arrays, Collections, Maps use JSON format (default for trunk)
- When false: All complex objects use toString() format
- Centralized format logic in tryConstructJson() method
- Maintains credential redaction for Maps before formatting

Trunk version defaults to JSON format. 5.0 branch patch will default
to false for backward compatibility.

Addresses CASSANDRA-20827 reviewer feedback for configurable output format.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/SettingsTable.java:
```diff
@@ -115,12 +119,31 @@ String getValue(Property prop)
 
+    private String tryConstructJson(Object o)
+    {
+        if (CassandraRelevantProperties.VIRTUAL_TABLE_COMPLEX_SETTINGS_FORMAT_JSON.getBoolean())
```
@mrktsm you can resolve the boolean value e.g. in the constructor and here it would be just simple `if (boolean)`, you do not need to resolve this every time. The value of  the property will be same during whole lifetime of the node.

mrktsm replied:
@smiklosovic Good point. Fixed the boolean resolution - moved it to the constructor like you suggested.

mrktsm committed:
Optimize boolean property resolution per reviewer feedback

Move CassandraRelevantProperties.VIRTUAL_TABLE_COMPLEX_SETTINGS_FORMAT_JSON.getBoolean()
resolution to constructor to avoid repeated calls in tryConstructJson().

Addresses Stefan Miklosovic's performance optimization suggestion.

mrktsm committed:
Add test for complex settings format property

- Add testComplexSettingsFormatProperty() to verify both unset and set system property behavior
- Test uses WithProperties pattern for property isolation
- Verifies JSON format when property is unset (default)
- Verifies toString format when property is set to false
- Covers both arrays (data_file_directories) and maps (seed_provider.parameters)
- Follows existing test patterns using check() method
